impl/src/main/java/org/jboss/cdi/tck/util/Assert.java

```diff
@@ -96,9 +108,49 @@ public static void assertTypeListMatches(List<? extends Type> types, Type... req
 
         List<Type> requiredTypeList = Arrays.asList(requiredTypes);
 
-        if (requiredTypes.length != types.size() || !types.containsAll(requiredTypeList)) {
+        if (requiredTypes.length != types.size() || !containsAllTypes(types, requiredTypeList)) {
             fail(String.format("List %s (%s) does not match array %s (%s)", types, types.size(), requiredTypeList, requiredTypeList.size()));
         }
     }
 
+    private static boolean containsAllTypes(Collection<? extends Type> existingTypes, Collection<? extends Type> requiredTypes) {
+        if (requiredTypes.isEmpty()) {
+            return false;
+        }
+        for (Type requiredType : requiredTypes) {
+            if (!containsType(existingTypes, requiredType)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private static boolean containsType(Collection<? extends Type> existingTypes, Type requiredType) {
+        if (existingTypes.isEmpty()) {
+            return false;
+        }
+        for (Type existingType : existingTypes) {
+            if (isTypesEqual(requiredType, existingType)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private static Boolean isTypesEqual(Type type, Type requiredType) {
+        return type.getTypeName().equals(requiredType.getTypeName());
```

Comment:
This seems rather imprecise? Also `Type.getTypeName()` by defaults delegates to `toString()`, so it's also quite fragile in presence of custom implementations of `Type`. IIUC, the original code relied on `Type.equals()`, which is likely to be much better. Of course it requires custom implementations of `Type` to interoperate correctly, which isn't that hard to do AFAICT. Here are some custom implementations of `Type` that (should :-) ) work correctly: https://github.com/weld/core/tree/master/weld-lite-extension-translator/src/main/java/org/jboss/weld/lite/extension/translator/util/reflection

Reply:
(It also requires the JDK implementations of `Type` to interoperate correctly, which AFAIK they do.)

Reply:
You are right, I wound a bug in our type implementation. Keeping only annotations changes.